### PR TITLE
fix(ui.open): some URLs fail on Windows

### PR DESCRIPTION
**Pull request by mrshmllow**
Problem:
On Windows, `explorer.exe` fails to open some URLs, for example:

    :lua vim.ui.open('https://devdocs.io/#q=lua%20lua_call')

https://github.com/neovim/neovim/pull/23401#issuecomment-1641015704

Solution:
Use rundll32 instead.

### CODE REVIEW

File: runtime/lua/vim/ui.lua [hunk 1]
```diff
@@ -136,7 +136,13 @@ function M.open(path)
   if vim.fn.has('mac') == 1 then
     cmd = { 'open', path }
   elseif vim.fn.has('win32') == 1 then
-    cmd = { 'explorer', path }
+    if vim.fn.executable('start') then
```
Comment (zeertzjq) [inline runtime/lua/vim/ui.lua:139]: I don't think `start` is an executable.

Comment (mrshmllow) [reply runtime/lua/vim/ui.lua]: I dont think start is even necessary

**Comment by mrshmllow** [comment]
Tested on windows 11 works on powershell & cmd

**Comment by adelarsq** [comment]
@mrshmllow  I had confirmed that with this fix vim.ui.open works as expected on Windows.

Will be necessary just an update on the docs for the new approach.

**Review by ObserverOfTime — COMMENTED** [review]
Is the check even necessary?

File: runtime/lua/vim/ui.lua [hunk 2]
```diff
@@ -136,7 +136,11 @@ function M.open(path)
   if vim.fn.has('mac') == 1 then
     cmd = { 'open', path }
   elseif vim.fn.has('win32') == 1 then
-    cmd = { 'explorer', path }
+    if vim.fn.executable('rundll32') then
+      cmd = { 'rundll32', 'url.dll,FileProtocolHandler', path }
+    else
+      return nil, 'vim.ui.open: rundll32 not on path.'
```
Comment (justinmk) [inline runtime/lua/vim/ui.lua:142]:
```suggestion
      return nil, 'vim.ui.open: rundll32 not found'
```

**Comment by mrshmllow** [comment]
it's kind of insane if rundll32 is not found but who knows with windows

**Comment by mrshmllow** [comment]
investigating that error, it seems to be because system() errors, something about `timeout`

**Comment by justinmk** [comment]
@adelarsq what does `:lua vim.print(vim.ui.open('non-existent-file'))` show on Windows?

**Review by theHamsta — COMMENTED** [review]
I think a `vimfn.executable('rundll') == 1` is missing since executable doesn't return a boolean

**Comment by justinmk** [comment]
Ah interesting, the problem is that `rundll32` silently fails. Did it open `non-existent-file` in any way, or just do nothing?

So we probably need to skip that test on Windows. Sigh.

**Comment by zeertzjq** [comment]
Doesn't that mean it never fails?

**Comment by justinmk** [comment]
> Doesn't that mean it never fails?

Right, in https://github.com/neovim/neovim/pull/24392#issuecomment-1642135214 by "silently fails" I mean that `rundll32` doesn't return a non-zero exit code, despite the fact that it didn't handle `non-existent-file` in a useful way.

**Comment by adelarsq** [comment]
I am already using the patch here. Working as expected for every link so far.

**Comment by mrshmllow** [comment]
maybe we manually check if the file exists or not..?

**Comment by zeertzjq** [comment]
Does the file really have to exist?

**Comment by justinmk** [comment]
File doesn't need to exist, it could be a URL.

> Does the api really need to guarrentee that the URI was opened properly? It is only used to report back to the LSP

? If open() fails the user sees an error message. That is useful feedback. But if rundll32 fails silently there's nothing we can do.

Thus the only thing blocking this PR... is fixing the test.

**Comment by adelarsq** [comment]
I did some research on this so the conclusion is that `rundll32` command itself does not have a built-in mechanism for returning status. So a test for this doesn't mean so much.

@justinmk Would be possible a mock around the behaviour for testing pourposes?

**Comment by justinmk** [comment]
Why would we mock? The test works on non-Windows. Just skip that part of the test in Windows, as I suggested above, by checking `is_os()` in the test.

**Comment by zeertzjq** [comment]
It only works on non-Windows by accident. If there is indeed an program that can open `nom-existent-file` then it won't work.

**Comment by justinmk** [comment]
> It only works on non-Windows by accident. If there is indeed an program that can open `nom-existent-file` then it won't work.

How is that an accident? There is no program on any OS, currently, that can open `non-existent-file`. It's just a Windows quirk that rundll32 doesn't report a failure even though it failed.

**Comment by justinmk** [comment]
Pushed a fix for the test.

**Comment by zeertzjq** [comment]
> None of these handlers on any OS, currently, can open `non-existent-file`. It's just a Windows quirk that rundll32 doesn't report a failure even though it failed.

For me `xdg-open` can do that.

**Comment by adelarsq** [comment]
> Why would we mock? The test works on non-Windows. Just skip that part of the test in Windows, as I suggested above, by checking is_os() in the test.

I did take a look on rundll to solve this issue and for now there is no good way to make it returns. So something like you did on this https://github.com/neovim/neovim/pull/24392/commits/83d4fdbf01f2592ca463b93ed0f484f1679f75ce I think is the best approach.

File: test/functional/lua/ui_spec.lua [hunk 3]
```diff
@@ -134,8 +135,11 @@ describe('vim.ui', function()
 
   describe('open()', function()
     it('validation', function()
-      if is_ci('github') then
-        matches('vim.ui.open: command failed %(%d%): { "[^"]+", "non%-existent%-file" }',
+      if is_os('win') or not is_ci('github') then
+        exec_lua[[vim.system = function() return { wait=function() return { code=3} end } end]]
+      end
+      if not is_os('bsd') then
```
Comment (zeertzjq) [inline test/functional/lua/ui_spec.lua:141]:
```suggestion
      else
```

Comment (zeertzjq) [reply test/functional/lua/ui_spec.lua:141]: Ah, I understand now.

**Comment by mrshmllow** [comment]
Thanks @justinmk

**Pull request merged**